Clean up doc comments for task APIs.

amartini51 opened the pull request:
Use only plain English words in abstracts, no symbol names, per
reference writing guidelines.

Use title case for headings.

Use spelling from Apple Style Guide for 'canceled' and 'canceling'.

Track TaskGroup async(...) renaming to addTask(...) in code examples.

Fill in the -Parameters: markup for methods and initializers that take
parameters, starting the sentence with an upper-case letter.

Add links to TSPL for conceptual discussion, re-using the prose from
elsewhere in the standard library reference for the cross reference.

Restore some prose from fe288c267c4c8f537b6b79567f228f8aa42755d4 in
places where cleanup and revision work to doc comments got overwritten
by API revision.

Restore some prose that was in the discussion of Task.Handle that was
deleted in commit 27bf2c97e5f3753ec31fc857f2b04e4a435ab4f9.

Reduce the number of places where we talk about cooperative cancellation
in detail, and let the individual properties and methods lean on the
longer discussion in the overview for Task.

Remove mention of error propagation from getResult(), which doesn't
throw and therefor can't propagate errors thrown by a task.

Remove mention of what the add(...) methods do if the operation throws
an error from the non-throwing TaskGroup type, since the operation there
can't ever be throwing.

Remove -Returns: markup from properties and from methods that don't
return a value.  Stored and computed properties *have* a value, which is
documented by their abstract and discussion sections; functions and
methods *return* a value.

Remove cross references when they refer to type that's part of the
declaration.  The rendered documentation already provides an affordance
to navigate to a type in that context.

Fixes rdar://81184794

amartini51 committed:
Clean up doc comments for task APIs.

Use only plain English words in abstracts, no symbol names, per
reference writing guidelines.

Use title case for headings.

Use spelling from Apple Style Guide for 'canceled' and 'canceling'.

Track TaskGroup async(...) renaming to addTask(...) in code examples.

Fill in the -Parameters: markup for methods and initializers that take
parameters, starting the sentence with an upper-case letter.

Add links to TSPL for conceptual discussion, re-using the prose from
elsewhere in the standard library reference for the cross reference.

Restore some prose from fe288c267c4c8f537b6b79567f228f8aa42755d4 in
places where cleanup and revision work to doc comments got overwritten
by API revision.

Restore some prose that was in the discussion of Task.Handle that was
deleted in commit 27bf2c97e5f3753ec31fc857f2b04e4a435ab4f9.

Reduce the number of places where we talk about cooperative cancellation
in detail, and let the individual properties and methods lean on the
longer discussion in the overview for Task.

Remove mention of error propagation from getResult(), which doesn't
throw and therefor can't propagate errors thrown by a task.

Remove mention of what the add(...) methods do if the operation throws
an error from the non-throwing TaskGroup type, since the operation there
can't ever be throwing.

Remove -Returns: markup from properties and from methods that don't
return a value.  Stored and computed properties *have* a value, which is
documented by their abstract and discussion sections; functions and
methods *return* a value.

Remove cross references when they refer to type that's part of the
declaration.  The rendered documentation already provides an affordance
to navigate to a type in that context.

Fixes <rdar://81184794>.

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -17,15 +17,23 @@ import Swift
+/// which can be either a detached task or a child task,
+/// `Task` also exposes class methods like `yield()` and `currentPriority`.
+/// Because all such functions are asynchronous,
+/// they're always invoked as part of an existing task.
```
`currentPriority` is not async so this needs some rewording.

currentPriority will try to get a priority from current task if called within one, and if not, it will try to make a best approximation of priority from the current thread priority.

Perhaps only mention yield here?

amartini51 replied:
Changed in 946c72cbc73 to just mention `yield`.

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -39,6 +47,12 @@ import Swift
 /// you don't generally interact with partial tasks directly.
 ///
+/// For information about the language-level concurrency model that `Task` is part of,
+/// see [Concurrency][concurrency] in *[The Swift Programming Language][tspl]*.
```
👍

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -39,6 +47,12 @@ import Swift
+/// see [Concurrency][concurrency] in *[The Swift Programming Language][tspl]*.
+///
+/// [concurrency]: https://docs.swift.org/swift-book/LanguageGuide/Concurrency.html
+/// [tspl]: https://docs.swift.org/swift-book/
```
👍

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -76,42 +90,34 @@ public struct Task<Success, Failure: Error>: Sendable {
-  /// If the task has not completed yet, its priority will be elevated to the
-  /// priority of the current task. Note that this may not be as effective as
-  /// creating the task with the "right" priority to in the first place.
+  /// If the task hasn't completed,
```
```suggestion
  /// If the task hasn't completed yet,
```

The yet somehow helps me understand what we mean here, wdyt?

amartini51 replied:
Sure, that's more specific. Added in 946c72cbc73.

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -76,42 +90,34 @@ public struct Task<Success, Failure: Error>: Sendable {
-  /// ### Cancellation
-  /// If the awaited on task gets cancelled externally the `get()` will throw
-  /// a cancellation error.
+  /// If the task throws an error, this method propogates that error.
```
```suggestion
  /// If the task throws an error, this computed property propogates that error.
```

I guess?

amartini51 replied:
Changed 'method' to 'property' here and in another spot in 946c72cbc73.

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -145,12 +156,10 @@ extension Task where Failure == Never {
+  /// Because this method is nonthrowing,
+  /// if the task checks for cancellation,
+  /// it needs to handle cancellation using an approach like returning `nil`
+  /// instead of throwing an error.
```
I suggest wording this differently, this isn't as much about this computed property being not throwing, but the entire `Task` type this is declared on is `Error == Never`, so it is more about:

> Since this Task is nonthrowing, this value cannot throw and if the task wanted to react to cancellation it could do so in some other way, e.g. by returning `nil` but it cannot throw.

?

amartini51 replied:
Rewrote in 946c72cbc73:

> Tasks that never throw an error can still check for cancellation,
> but they need to use an approach like returning `nil`
> instead of throwing an error.

I think this version is a little more direct too.

ktoso replied:
Sounds good :)

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -122,14 +128,19 @@ extension Task {
   }
 
-  /// Attempt to cancel the task.
+  /// Indicate that the task should stop.
```
I don't think "stop" is a safe word to use here, it makes one thing we actively "stop tasks from executing" which isn't the case. It's a bit repetetive but I think we must stick to cancellation wording here, and not introduce new verbs like stopping etc.

amartini51 replied:
Maybe we can brainstorm verbs?  The guidelines for writing reference strongly discourage repeating the API symbol name in the abstract (this first line of docs), to avoid the problem of saying things like:

ktoso replied:
I see what you mean with it then just repeating the API name... I don't have any other great ideas though :-( Stops makes me pretty worried, we never "stop" anything, the task may continue to run forever for all we know hmm...

You're right on "attempts" that'd be potentially misleading.

ktoso replied:
Maybe... 

> Puts the task in cancelled state, for which it may or may not check and act upon.

?

amartini51 replied:
Thanks.  How about "Requests that the task should cancel its work."?  Or "Indicates that the task should stop running."?

The fact that a task might ignore cancellation doesn't need to be in the abstract; we discuss it in the paragraphs below.

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -439,22 +445,29 @@ func taskCreateFlags(
-  /// inherits the priority and actor context of the caller, so the `operation`
-  /// is treated more like an asynchronous extension to the synchronous
-  /// operation.
+  /// You need to keep a reference to the detached task
```
```suggestion
  /// You need to keep a reference to the task
```

this isn't detached, so just "to the task" is okey; or "to the unstructured task"

amartini51 replied:
Changed here and on the other init below in 946c72cbc73.

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -17,15 +17,23 @@ import Swift
+/// It's not a programming error to discard a reference to a task
+/// without waiting for that task to finish or canceling it.
+/// A task runs whether or not you keep a reference to it.
+/// However, if you discard a task's handle, you give up the ability
```
You seem to move wording to use "task reference" so perhaps also remove the "handle" word here?

```suggestion
/// However, if you discard a task reference, you give up the ability
```

amartini51 replied:
Changed in 946c72cbc73, thanks.

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -738,7 +744,10 @@ public struct UnsafeCurrentTask {
+  /// A Boolean value that indicates whether the current task was canceled.
+  ///
+  /// After the value of this property is `true`, it remains `true` indefinitely.
+  /// There is no way to uncancel the operation.
```
```suggestion
  /// There is no way to uncancel a task.
```

?

amartini51 replied:
Changed in 946c72cbc73.

ktoso commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -787,6 +788,27 @@ extension ThrowingTaskGroup: AsyncSequence {
+  ///     }
+  ///     
+  ///     // Assuming the child tasks complete in order, this prints "2"
+  ///     for try await r in group { print(r) }
```
Good example, thanks!

ktoso reviewed (approved):
A few minor things, looks good -- thank you!

amartini51 committed:
Incorporate tech review corrections.

Feedback from @ktoso.

amartini51 committed:
Match throwing and nonthrowing next() docs.

amartini51 commented:
@swift-ci Please test

amartini51 committed:
Apply changes from editorial.

Changes from <rdar://81715064>

amartini51 committed:
Back out edits that are out of scope.

This includes:

- Edits to /// comments for non-public symbols
- Edits to // comments (not documentation)
- Edits to warnings and other strings in code

ktoso commented:
Thanks Alex, nice edits! 

I pondered a bit about the cancel docs, let me know what you think but the PR is LGTM 👍

amartini51 committed:
Cross-apply edit: TSPL isn't in italics.

amartini51 committed:
Revise editorial changes.

amartini51 committed:
Use "that" for restrictive clause, per Apple style.

amartini51 committed:
Remove invalid markup.

Fixes <rdar://82053644>.

amartini51 commented:
@swift-ci Please test.

tbkka commented on stdlib/public/Concurrency/TaskCancellation.swift:
```diff
@@ -34,8 +34,8 @@ public func withTaskCancellationHandler<T>(
-  // unconditionally add the cancellation record to the task.
-  // if the task was already cancelled, it will be executed right away.
+  // Unconditionally add the cancellation record to the task.
+  // If the task was already canceled, it executes immediately.
```
"If the task was already canceled, it executes immediately."

This makes it sound like "the task executes immediately", which I presume is not the case for a task that was already canceled.

amartini51 replied:
Nice catch - this is a `//` comment, not part of the documentation, so should not be changed by these edits.  I did a visual check for these for commit 6a79590ca67fef5c9d05debd253588437e96e99d; and I'll run my usual `grep` script to check for others now.

amartini51 replied:
Reverted in a0e58c68860.  @ktoso or @DougGregor do you want to work with @tbkka to improve this internal comment as needed?

amartini51 committed:
Back out a non-documentation comment change.

Thanks, @tbkka!

Confirmed this was the only one by running:

    git diff github/release/5.5-08092021...HEAD \
    | grep '^+\|^-' \
    | grep -v '^+++\|^---' \
    | grep -v '^- *///\|^+ *///'

amartini51 committed:
Correct spelling errors and doubled word.

amartini51 commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -397,6 +388,8 @@ public struct TaskGroup<ChildTaskResult> {
   /// If the task that's currently running this group is canceled,
   /// the group is also implicitly canceled,
   /// which is also reflected in this property's value.
+  /// ◊TR: Why do we have two implementations of this method?
+  /// ◊TR: What's the difference between them?
```
@ktoso or @DougGregor can you comment?

amartini51 commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -376,13 +364,16 @@ public struct TaskGroup<ChildTaskResult> {
   ///
-  /// This function may be called even from within child (or any other) tasks,
-  /// and causes the group to be canceled.
+  /// There are no restrictions on where you can call this method.
+  /// Code inside a child task or even another task can cancel a group.
+  /// ◊TR: Is this rewrite too strong?
```
@ktoso or @DougGregor can you comment?

amartini51 commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -656,18 +658,21 @@ public struct ThrowingTaskGroup<ChildTaskResult, Failure: Error> {
   ///
-  /// This function may be called even from within child (or any other) tasks,
-  /// and causes the group to be canceled.
+  /// There are no restrictions on where you can call this method.
+  /// Code inside a child task or even another task can cancel a group.
+  /// ◊TR: Is this rewrite too strong?
```
@ktoso or @DougGregor can you comment?

amartini51 commented on stdlib/public/Concurrency/TaskGroup.swift:
```diff
@@ -679,6 +684,8 @@ public struct ThrowingTaskGroup<ChildTaskResult, Failure: Error> {
   /// If the task that's currently running this group is canceled,
   /// the group is also implicitly canceled,
   /// which is also reflected in this property's value.
+  /// ◊TR: Why do we have two implementations of this method?
+  /// ◊TR: What's the difference between them?
```
@ktoso or @DougGregor can you comment?

amartini51 commented on stdlib/public/Concurrency/TaskSleep.swift:
```diff
@@ -14,10 +14,12 @@ import Swift
-  /// Suspends the current task for _at least_ the given duration
+  /// Suspends the current task for at least the given duration
   /// in nanoseconds.
+  /// ◊TR: Why do we have both sleep(_:) and sleep(nanoseconds:) -- what's the difference between them?
+  /// ◊TR: It looks like only sleep(nanoseconds:) throws.
```
@ktoso or @DougGregor can you comment?

tbkka replied:
I thought we were going to deprecate `sleep(_:)` so that we could later bring it back when we had a legitimate `Duration` type.  If the argument is just a raw number, then denoting the units as `sleep(nanoseconds:)` is much clearer.

amartini51 replied:
That sounds right to me, but I don't see an `@availability` annotation on `sleep(_:)` that marks it as deprecated here or on the [release branch](https://github.com/apple/swift/blob/release/5.5-08092021/stdlib/public/Concurrency/TaskSleep.swift#L15).  It sounds like there should be one?

amartini51 committed:
Remove markers for tech review queries.

amartini51 commented:
@swift-ci Please test.

amartini51 commented:
@swift-ci Please test macOS platform.
